src/plugins/discover/public/application/main/discover_main_route.tsx

```diff
@@ -252,6 +257,10 @@ export function DiscoverMainRoute({
     [loadSavedSearch]
   );
 
+  const onESQLNavigationComplete = useCallback(async () => {
+    resetStateContainer();
```

Comment:
This is to address the bug that when you are in Discover and click on the empty prompt "Try ES|QL" the empty prompt doesn't disappear. This is because even though the URL state was updated with the new es|ql, query, discover didn't pick up the update and didn't remove the prompt. The problem is that the URL/state synching logic wasn't initialized at this point as it is initialized inside the discover app component.

To workaround I landed on this callback approach that allows to do something after the user clicks the "Try ES|QL" and the URL has changed. In this case, we reinitialize the state container so that it picks up the es|ql query as the initial state. 

Another simpler alternative I tried was to do a full-page reload, but it seemed too slow.